### PR TITLE
refactor(lint): refactor NoParametersOnlyUsedInRecursion

### PR DESCRIPTION
**Pull request by matanshavit**
## Summary
Addresses comments on #7926, which was merged before all comments were addressed.
This should be a refactor only, not change any behavior.

## AI assistance notice
I use Claude Code to write most of the code and comments in this pull request. As the human author, I take full responsibility for its contents

## Test Plan
Existing unit and snapshot tests pass. No behavior should be changed.

## Docs
- added docstrings to helper functions
- Corrected incorrect, misleading, and unneeded comments

### CODE REVIEW

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 1]
```diff
@@ -294,29 +281,16 @@ fn get_arrow_function_name(
 
         // Stop searching if we hit a function boundary
         // (prevents extracting wrong name from outer scope)
-        if is_function_like(&ancestor) {
+        if AnyFunctionLike::cast_ref(&ancestor).is_some() {
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: You can use `can_cast(ancestor.kind())`

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 2]
```diff
@@ -335,20 +309,22 @@ fn is_function_signature(parent_function: &AnyJsParameterParentFunction) -> bool
     )
 }
 
-fn is_recursive_call(call: &JsCallExpression, function_name: Option<&TokenText>) -> bool {
+/// Checks if a call expression is a recursive call to the current function.
+/// Handles direct calls (`foo()`), method calls (`this.foo()`), and computed members (`this["foo"]()`).
+/// Uses a conservative approach to avoid false positives.
+fn is_recursive_call(call: &JsCallExpression, function_name: &TokenText) -> bool {
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: This should return `Option<bool>`, so you can simplify a lot the internal code by removing let-else chain and using the try operator

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs (hunk 1 above)
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: Thanks, changed

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 3]
```diff
@@ -335,20 +309,22 @@ fn is_function_signature(parent_function: &AnyJsParameterParentFunction) -> bool
...
     let expr = callee.omit_parentheses();
 
     // Simple identifier: foo()
     if let Some(ref_id) = expr.as_js_reference_identifier() {
-        return ref_id.name().ok().is_some_and(|n| n.text() == name.text());
+        return ref_id
+            .name()
+            .ok()
+            .is_some_and(|n| n.text() == function_name.text());
     }
```
[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: I think these should actually be different, the way they currently are, that trimming isn't necessary here and we should just directly compare to the token. I suggest we reject this comment and keep the code change a refactor, not a functional change.

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs (hunk 2 above)
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: I understand, this is like the other functions that should return Option<bool>. Changed.

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: > that trimming isn't necessary here and we should just directly compare to the token

I'm not sure what you mean, but the bot is onto something. Trivia are attached to tokens, so whitespaces and comments actually can interfere

[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: Oh ok, that does make sense then. I was looking at function_name and kept thinking it is already trimmed. But that is the exact issue if the left expression is not trimmed.

I made the change to line 323, but on line 355, `inner_string_text` already uses `token_text_trimmed`, so I just added a comment.

[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: I guess the confusion I had is the function_name is already trimmed when generated, let me try to make the code clearer

[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: Ok changed. How about we just compare instances of TokenText instead of trying to get the strong text to compare?
example:
```
name.token_text_trimmed() == *function_name
```

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 4]
```diff
@@ -414,8 +375,8 @@ fn is_reference_in_recursive_call(
         // Check if this is a call expression
         if let Some(call_expr) = JsCallExpression::cast_ref(&node) {
             // Check if this call is recursive AND uses our parameter
-            if is_recursive_call_with_param_usage(&call_expr, function_name, param_name) {
-                return true;
+            if is_recursive_call_with_param_usage(&call_expr, function_name, param_name)? {
+                return Some(true);
             }
         }
```
[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs:472]: Ok done

**Comment by matanshavit** [comment]
re: coderabbitai nitpick - seems good! I like consistency

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]: Yeah that should work 👍

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 5]
```diff
@@ -124,27 +122,21 @@ impl Rule for NoParametersOnlyUsedInRecursion {
...
+        for reference in binding.all_references(model) {
             if is_reference_in_recursive_call(
                 &reference,
-                function_name.as_ref(),
+                &function_name,
                 &parent_function,
                 name_text,
-            ) {
+            )
+            .unwrap_or(false)
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs]:
```suggestion
            .unwrap_or_default()
```

**Comment by ematipico** [comment]
@matanshavit can you please add a changeset?

**Comment by dyc3** [comment]
@ematipico this is just an internal refactor. none of the tests changed. this doesn't need a changeset, unless I missed something?

File: crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs [hunk 6]
```diff
@@ -443,104 +405,86 @@ fn is_function_boundary(
...
+            to_check.push(member_expr.object().ok()?);
         }
 
         // Any other expression - not safe to trace
         // Just continue to next item in worklist
     }
 
     // Didn't find the parameter anywhere
-    false
+    Some(false)
```
[the remark this answers is not in the document]
Comment (matanshavit) [reply crates/biome_js_analyze/src/lint/nursery/no_parameters_only_used_in_recursion.rs:575]: I thought about this and I agree, the linting will be more robust if it can still detect this rule being broken even if it is part of a larger malformed expression. I made a commit for that as well

**Comment by matanshavit** [comment]
It's possible some edge cases changed. It might be a lot of new test code to test for every single difference of semantics. It seems excessive to do that at this point, but correct me if we need more compound expression testing (like where a rule violation is contained in a larger expression that is malformed).
I can create a changeset regardless, just in case.

File: .changeset/flat-results-happen.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+"@biomejs/biome": patch
+---
+
+refactor NoParametersOnlyUsedInRecursion and refine to accurately lint in valid subexpressions in malformed expressions
```
Comment (dyc3) [inline .changeset/flat-results-happen.md]: should be past tense, and it's not really clear to users what changed and how it affects them. does it fix something?

**Comment by ematipico** [comment]
Ah sorry. The PR description is basically wrong and points to the wrong PR

**Review by ematipico — APPROVED** [review]

**Comment by matanshavit** [comment]
My bad on pr description, linked correct old pr

**Comment by ematipico** [comment]
Happy to merge once the conflicts are resolved
